core/src/sumeragi/network_topology.rs

```diff
@@ -234,4 +356,202 @@ mod tests {
+    }
+
+    #[test]
+    fn roles() {
+        let peers = peers![0, 1, 2, 3, 4, 5, 6];
+        let topology = Topology::new(peers.clone());
+        let expected_roles = [
+            Role::Leader,
+            Role::ValidatingPeer,
+            Role::ValidatingPeer,
+            Role::ValidatingPeer,
+            Role::ProxyTail,
+            Role::ObservingPeer,
+            Role::ObservingPeer,
+        ];
+
+        for ((i, peer), expected_role) in (0..).zip(peers).zip(expected_roles) {
+            let actual_role = topology.role(&peer);
+            assert_eq!(
+                actual_role, expected_role,
+                "Role detection failed for peer with index: {i}"
+            );
+        }
+    }
+
+    #[test]
+    fn proxy_tail() {
+        let peers = peers![0, 1, 2, 3, 4, 5, 6];
+        let topology = Topology::new(peers.clone());
+
+        assert_eq!(topology.proxy_tail(), &peers[4]);
+    }
+
+    #[test]
+    fn leader() {
+        let peers = peers![0, 1, 2, 3, 4, 5, 6];
+        let topology = Topology::new(peers.clone());
+
+        assert_eq!(topology.leader(), &peers[0]);
+    }
+
+    #[test]
+    fn signature_verification_ok() {
+        let key_pairs =
+            core::iter::repeat_with(|| KeyPair::generate().expect("Failed to generate key pair"))
+                .take(7)
+                .collect::<Vec<_>>();
+        let mut key_pairs_iter = key_pairs.iter();
+        let peers = peers![0, 1, 2, 3, 4, 5, 6: key_pairs_iter];
+        let topology = Topology::new(peers);
+
+        let dummy = "value to sign";
+        let mut signatures = key_pairs
+            .iter()
+            .map(|key_pair| SignatureOf::new(key_pair.clone(), &dummy).expect("Failed to sign"))
+            .collect::<Result<SignaturesOf<_>, _>>()
+            .expect("Failed to create `SignaturesOf`");
+
+        assert_eq!(
+            topology.verify_signatures(&mut signatures, HashOf::new(&dummy)),
+            Ok(())
+        );
+    }
+
+    #[test]
+    fn signature_verification_consensus_not_required_ok() {
+        let key_pairs =
+            core::iter::repeat_with(|| KeyPair::generate().expect("Failed to generate key pair"))
+                .take(3)
+                .collect::<Vec<_>>();
+        let mut key_pairs_iter = key_pairs.iter();
+        let peers = peers![0, 1, 2: key_pairs_iter];
+        let topology = Topology::new(peers);
+
+        let dummy = "value to sign";
+        let mut signatures = key_pairs
+            .iter()
+            .enumerate()
+            .filter(|(i, _)| *i == 2) // Retain only last signature
+            .map(|(_, key_pair)| SignatureOf::new(key_pair.clone(), &dummy).expect("Failed to sign"))
+            .collect::<Result<SignaturesOf<_>, _>>()
+            .expect("Failed to create `SignaturesOf`");
+
+        let result = topology.verify_signatures(&mut signatures, HashOf::new(&dummy));
+        assert_eq!(result, Ok(()))
```

Comment:
```suggestion
        assert!(topology.verify_signatures(&mut signatures, HashOf::new(&dummy)).is_ok())
```

Reply:
I would like to keep is this way because this way we can see `rhs` and what exact error we received.